File: prototype/OptimalControl/LowThrust/@DecisionVector/DecisionVector.m
classdef DecisionVector < handle
    
    %  Author: S. Hughes.  steven.p.hughes@nasa.gov
    %
    %  This class manages parts of a decision vector allowing you to create
    %   a decision vector, extract parts, or set parts of the vector.
    %  The decsion vector is organized according to Eq. 41 of GPOPS II
    %  journal article.
    %  Z = [V' W' Q' t' s']
    %  where
    %  v = [X1 Y1 Z1 VX1 VY1 VZ1 M1]
    %      [X2 Y2 Z2 VX2 VY2 VZ2 M2]
    %      [XN YN ZN VXN VYN VZN M1]
    %  and
    %  w = [Ux1 Uy1 Uz1]
    %      [Ux2 Uy2 Uz2]
    %      [UxN UyN UzN]
    %
    
    properties  (SetAccess = 'protected')
        
        %  Properties for continuous optimal control problem
        numStates
        numControls
        numIntegralParams
        numStaticParams
        
        %  Properties of the discretization of the state and controls
        numStateMeshPoints
        numControlMeshPoints
        numStateStagePoints;
        numControlStagePoints;
        numStatePoints
        numControlPoints
        numStatePointsPerMesh
        numControlPointsPerMesh
        
        %  Properties of the complete decision variable vector
        lengthDecVector
        decisionVector
        numStateParams
        numControlParams
        
        %  Vector Chunking Ideces
        stateStartIdx
        stateStopIdx
        controlStartIdx
        controlStopIdx
        integralStartIdx
        integralStopIdx
        timeStartIdx
        timeStopIdx
        staticStartIdx
        staticStopIdx
        
    end
    
    methods
        
        %% Initialize the decision vector. Set properties. Validate values.
        function Initialize(obj,numStates, ...
                numControls,numIntegrals,numStaticParams,...
                numStateMeshPoints,numControlMeshPoints,...
                numStateStagePoints,numControlStagePoints)
            
            %  Check that inputs are valid
            if  numStates <= 0
                errorMsg = 'numStates must be > 0';
                errorLoc  = 'DecisionVector:ConfigureDecisionVector';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            %  if  numControls <= 0
            %     errorMsg = 'numControls must be > 0';
            %     errorLoc  = 'DecisionVector:ConfigureDecisionVector';
            %     ME = MException(errorLoc,errorMsg);
            %     throw(ME);
            %  end
            if  numStateMeshPoints <= 0
                errorMsg = 'numStateMeshPoints must be > 0';
                errorLoc  = 'DecisionVector:ConfigureDecisionVector';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            if  numControlMeshPoints <= 0
                errorMsg = 'numControlMeshPoints must be > 0';
                errorLoc  = 'DecisionVector:ConfigureDecisionVector';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            
            %  Set optimal control problem sizes
            obj.numStates         = numStates;
            obj.numControls       = numControls;
            obj.numIntegralParams = numIntegrals;
            obj.numStaticParams   = numStaticParams;
            
            %  Set discretization properites
            obj.numStateMeshPoints    = numStateMeshPoints;
            obj.numControlMeshPoints  = numControlMeshPoints;
            obj.numStateStagePoints   = numStateStagePoints;
            obj.numControlStagePoints = numControlStagePoints;
            
            if obj.numStateStagePoints == 0
                obj.numStatePoints = numStateMeshPoints;
            else
                obj.numStatePoints        = numStateMeshPoints*(1 +...
                    numStateStagePoints) - 1;
            end
            if obj.numControlStagePoints == 0
                obj.numControlPoints = numControlMeshPoints;
            else
            obj.numControlPoints        = numControlMeshPoints*(1 +...
                numControlStagePoints) - 1;
            end
            obj.numStatePointsPerMesh = 1 + numStateStagePoints;
            obj.numControlPointsPerMesh = 1 + numControlStagePoints;
            obj.numStateParams    = numStates*obj.numStatePoints;
            obj.numControlParams  = numControls*obj.numControlPoints;
            obj.lengthDecVector   = obj.numStates*obj.numStatePoints...
                + obj.numControls*obj.numControlPoints + ...
                obj.numIntegralParams + obj.numStaticParams + 2;
            
            %  Compute the start and stop indeces of chunks
            obj.stateStartIdx    = 1;
            obj.stateStopIdx     = obj.numStateParams;
            obj.controlStartIdx  = obj.numStateParams + 1;
            obj.controlStopIdx   = obj.controlStartIdx + ...
                obj.numControlParams - 1;
            obj.integralStartIdx = obj.controlStopIdx + 1;
            obj.integralStopIdx  = obj.integralStartIdx + ...
                obj.numIntegralParams - 1;
            obj.timeStartIdx     = obj.integralStopIdx + 1;
            obj.timeStopIdx      = obj.timeStartIdx +  1;
            obj.staticStartIdx   = obj.timeStopIdx + 1;
            obj.staticStopIdx    = obj.staticStartIdx + ...
                obj.numStaticParams - 1;
            %
        end
        
        function SetDecisionVector(obj,decisionVector)
            
            %  Check that decision vector is the required length
            [rowSize,colSize] = size(decisionVector);
            if  rowSize ~= obj.lengthDecVector || colSize ~=1
                errorMsg = 'State vector is not valid length';
                errorLoc  = 'DecisionVector:SetDecisionVector';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            
            %  Set the decision vector
            obj.decisionVector    = decisionVector;
            
        end
        
        %  Given a state array, insert it in the decision vector
        function SetStateArray(obj,stateArray)
            
            [row,col] = size(stateArray);
            if  row ~= obj.numStatePoints || col ~= obj.numStates
                errorMsg = 'State Array is not valid dimension';
                errorLoc  = 'DecisionVector:SetStateArray';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            
            stateVector = reshape(stateArray,...
                obj.numStateParams,1);
            obj.decisionVector(obj.stateStartIdx:obj.stateStopIdx,1)...
                = stateVector;
            
        end
        
        %  Given a state vector, insert it in the decision vector
        function SetStateVector(obj,stateVector)
            
            [row,col] = size(stateVector);
            if  row ~= obj.numStateParams || col ~= 1
                errorMsg = 'State Vector is not valid dimension';
                errorLoc  = 'DecisionVector:SetStateVector';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            
            obj.decisionVector(obj.stateStartIdx:obj.stateStopIdx,1)...
                = stateVector;
            
        end
        
        %  Given a control array, insert it in the decision vector
        function SetControlArray(obj,controlArray)
            
            [row,col] = size(controlArray);
            if  row ~= obj.numControlPoints || col ~= obj.numControls
                errorMsg = 'Control Array is not valid dimension';
                errorLoc  = 'DecisionVector:SetControlArray';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            
            controlVector = reshape(controlArray,obj.numControlParams,1);
            obj.decisionVector(obj.controlStartIdx:obj.controlStopIdx,1)...
                = controlVector;
            
        end
        
        %  Given a control vector, insert it in the decision vector
        function SetControlVector(obj,controlVector)
            
            [row,col] = size(controlVector);
            if  row ~= obj.numControlParams || col ~= 1
                errorMsg = 'Control Vector is not valid dimension';
                errorLoc = 'DecisionVector:SetControlVector';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            
            obj.decisionVector(obj.controlStartIdx:obj.controlStopIdx,1)...
                = controlVector;
            
        end
        
        %  Given a time vector, insert it in the decision vector
        function SetTimeVector(obj,timeVector)
            
            [row,col] = size(timeVector);
            if  row ~= 2 || col ~= 1
                errorMsg = 'Time Vector is not valid dimension';
                errorLoc = 'DecisionVector:SetTimeVector';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            obj.decisionVector(obj.timeStartIdx:obj.timeStopIdx,1)...
                = timeVector;
            
        end
        
        %  Given a static vector, insert it in the decision vector
        function SetStaticVector(obj,staticVector)
            
            [row,col] = size(staticVector);
            if  row ~= obj.numStaticParams || col ~= 1
                errorMsg = 'Static Vector is not valid dimension';
                errorLoc = 'DecisionVector:SetStaticVector';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            obj.decisionVector(obj.staticStartIdx:obj.staticStopIdx,1)...
                = staticVector;
            
        end
        
        %  Given an integral vector, insert it in the decision vector
        function SetIntegralVector(obj,integralVector)
            
            [row,col] = size(integralVector);
            if  row ~= obj.numIntegralParams || col ~= 1
                errorMsg = 'Integral Vector is not valid dimension';
                errorLoc = 'DecisionVector:SetIntegralVector';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            obj.decisionVector(obj.integralStartIdx:...
                obj.integralStopIdx,1)= integralVector;
            
        end
        
        %  Extract part of decision vector that contains states
        function stateSubVector = GetStateSubVector(obj)
            stateSubVector = obj.decisionVector(1:obj.stateStopIdx);
        end
        
        %  Extract part of decision vector that contains controls
        function controlSubVector = GetControlSubVector(obj)
            controlSubVector = obj.decisionVector(obj.controlStartIdx: ...
                obj.controlStopIdx);
        end
        
        %  Extract part of decision vector that contains integrals
        function integralSubVector = GetIntegralSubVector(obj)
            integralSubVector = obj.decisionVector(...
                obj.integralStartIdx:obj.integralStopIdx);
        end
        
        %  Extract part of decision vector that contains static params
        function staticSubVector = GetStaticSubVector(obj)
            staticSubVector = obj.decisionVector(...
                obj.staticStartIdx:obj.staticStopIdx);
        end
        
        %  Extract part of decision vector that contains times
        function timeSubVector = GetTimeSubVector(obj)
            timeSubVector = obj.decisionVector(...
                obj.timeStartIdx:obj.timeStopIdx);
        end
        
        %  Extract the state array
        function stateArray = GetStateArray(obj)
            
            %  stateArray is numStateParams x numStates
            %
            %  stateArray = [X1 Y1 Z1 VX1 VY1 VZ1 M1]
            %               [X2 Y2 Z2 VX2 VY2 VZ2 M2]
            %               [XN YN ZN VXN VYN VZN M1]
            
            stateArray = reshape(obj.decisionVector(obj.stateStartIdx:...
                obj.stateStopIdx),obj.numStatePoints,obj.numStates);
            
        end
        
        %  Extract the state array
        function controlArray = GetControlArray(obj)
            
            %  controlArray is numControlMeshPoints x numControls
            %
            %  Build a dummy W matrix in Eq. 41.
            %  W = [Ux1 Uy1 Uz1]
            %      [Ux2 Uy2 Uz2]
            %      [UxN UyN UzN]
            
            controlArray = reshape(obj.decisionVector...
                (obj.controlStartIdx:obj.controlStopIdx),...
                obj.numControlPoints,obj.numControls);
            
        end
        
        %  Get the first state vector
        function stateVec = GetFirstStateVector(obj)
            stateVec = GetStateAtMeshPoint(obj,1);
        end
        
        %  Get the last state vector
        function stateVec = GetLastStateVector(obj)
            stateVec = GetStateAtMeshPoint(obj,obj.numStateMeshPoints);
        end
        
        %  Get the first control vector
        function stateVec = GetFirstControlVector(obj)
            stateVec = GetControlAtMeshPoint(obj,1);
        end
        
        %  Get the last control vector
        function stateVec = GetLastControlVector(obj)
            stateVec = GetControlAtMeshPoint(obj,obj.numControlPoints);
        end
        
        %  Get first time
        function time = GetFirstTime(obj)
            time = obj.decisionVector(obj.timeStartIdx);
        end
        
        %  Get last time
        function time = GetLastTime(obj)
            time = obj.decisionVector(obj.timeStopIdx);
        end
        
        %  Extract state vector at given mesh point
        function stateVec = GetStateAtMeshPoint(obj,meshIdx,stageIdx)
            % Check that state vector contains the mesh point
            if  meshIdx <= 0 || meshIdx > obj.numStatePoints
                errorMsg = ['meshIdx must be >= 0 and <= ' ...
                    'numStateMeshPoints'];
                errorLoc  = 'DecisionVector:GetStateAtMeshPoint';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            
            %  Set state to zero if this is a a mesh point
            if nargin == 2
                stageIdx = 0;
            end
            
            %  Compute indeces of state vector components the extract the
            %  state vector
            startIdx = (obj.stateStartIdx - 1) + ...
                (meshIdx-1)*obj.numStatePointsPerMesh + ...
                1 + stageIdx;
            step     = obj.numStatePoints;
            stopIdx = startIdx + step*(obj.numStates-1);
            vecIndeces = startIdx:step:stopIdx;
            stateVec = obj.decisionVector(vecIndeces);
        end
        
        %  Extract control vector at given mesh point
        function controlVec = GetControlAtMeshPoint(obj,meshIdx,stageIdx)
            % Check that state vector contains the mesh point
            if  meshIdx <= 0 || meshIdx > obj.numControlPoints
                errorMsg = ['meshIdx must be >= 0 and <= ' ...
                    ' numControlMeshPoints'];
                errorLoc  = 'DecisionVector:GetControlAtMeshPoint';
                ME = MException(errorLoc,errorMsg);
                throw(ME);
            end
            
            %  Set state to zero if this is a mesh point
            if nargin < 3
                stageIdx = 0;
            end
            
            %  Compute indeces of control vector components the extract the
            %  control vector
            startIdx = (obj.controlStartIdx - 1) + ...
                (meshIdx-1)*obj.numControlPointsPerMesh + ...
                1 + stageIdx;
            step     = obj.numControlPoints;
            stopIdx = startIdx + step*(obj.numControls-1);
            vecIndeces = startIdx:step:stopIdx;
            controlVec = obj.decisionVector(vecIndeces); 
        end
        
    end
    
end

File: src/csalt/src/collutils/DecisionVector.cpp
//------------------------------------------------------------------------------
//                              DecisionVector
//------------------------------------------------------------------------------
// GMAT: General Mission Analysis Tool
//
// Copyright (c) 2002 - 2020 United States Government as represented by the
// Administrator of The National Aeronautics and Space Administration.
// All Other Rights Reserved.
//
// Author: Wendy Shoan / NASA
// Created: 2015.06.17
//
/**
 * From original MATLAB prototype:
 *  Author: S. Hughes.  steven.p.hughes@nasa.gov
 *
 *  This class manages parts of a decision vector allowing you to create
 *  a decision vector, extract parts, or set parts of the vector.
 *
 */
//------------------------------------------------------------------------------

#include "DecisionVector.hpp"
#include "LowThrustException.hpp"
#include "MessageInterface.hpp"

//#define DEBUG_SET
//#define DEBUG_GET
//#define DEBUG_INIT
//#define DEBUG_DEC_VEC
//#define DEBUG_DEC_VEC_STATE

//------------------------------------------------------------------------------
// static data
//------------------------------------------------------------------------------
// none at this time

//------------------------------------------------------------------------------
// public methods
//------------------------------------------------------------------------------

//------------------------------------------------------------------------------
// default constructor
//------------------------------------------------------------------------------
DecisionVector::DecisionVector() :
   numStateVars            (0),
   numControlVars          (0),
   numStateMeshPoints      (0),
   numControlMeshPoints    (0),
   numStateStagePoints     (0),
   numControlStagePoints   (0),
   numStatePoints          (0),
   numControlPoints        (0),
   numStatePointsPerMesh   (0),
   numControlPointsPerMesh (0),
   numDecisionParams       (0),
   numStateParams          (0),
   numControlParams        (0),
   numIntegralParams       (0),
   numStaticParams         (0),
   integralStartIdx        (0),
   integralStopIdx         (0),
   timeStartIdx            (0),
   timeStopIdx             (0),
   staticStartIdx          (0),
   staticStopIdx           (0)
{
   // by default decisionVector is unsized
}


//------------------------------------------------------------------------------
// copy constructor
//------------------------------------------------------------------------------
DecisionVector::DecisionVector(const DecisionVector &copy) :
   numStateVars            (copy.numStateVars),
   numControlVars          (copy.numControlVars),
   numStateMeshPoints      (copy.numStateMeshPoints),
   numControlMeshPoints    (copy.numControlMeshPoints),
   numStateStagePoints     (copy.numStateStagePoints),
   numControlStagePoints   (copy.numControlStagePoints),
   numStatePoints          (copy.numStatePoints),
   numControlPoints        (copy.numControlPoints),
   numStatePointsPerMesh   (copy.numStatePointsPerMesh),
   numControlPointsPerMesh (copy.numControlPointsPerMesh),
   numDecisionParams       (copy.numDecisionParams),
   numStateParams          (copy.numStateParams),
   numControlParams        (copy.numControlParams),
   numIntegralParams       (copy.numIntegralParams),
   numStaticParams         (copy.numStaticParams),
   integralStartIdx        (copy.integralStartIdx),
   integralStopIdx         (copy.integralStopIdx),
   timeStartIdx            (copy.timeStartIdx),
   timeStopIdx             (copy.timeStopIdx),
   staticStartIdx          (copy.staticStartIdx),
   staticStopIdx           (copy.staticStopIdx)
{
   Integer sz = (Integer) copy.decisionVector.GetSize();
   decisionVector.SetSize(sz);
   for (Integer ii = 0; ii < sz; ii++)
      decisionVector[ii] = copy.decisionVector[ii];
}

//------------------------------------------------------------------------------
// operator=
//------------------------------------------------------------------------------
DecisionVector& DecisionVector::operator=(const DecisionVector &copy)
{
   
   if (&copy == this)
      return *this;
   
   numStateVars             = copy.numStateVars;
   numControlVars           = copy.numControlVars;
   numStateMeshPoints       = copy.numStateMeshPoints;
   numControlMeshPoints     = copy.numControlMeshPoints;
   numStateStagePoints      = copy.numStateStagePoints;
   numControlStagePoints    = copy.numControlStagePoints;
   numStatePoints           = copy.numStatePoints;
   numControlPoints         = copy.numControlPoints;
   numStatePointsPerMesh    = copy.numStatePointsPerMesh;
   numControlPointsPerMesh  = copy.numControlPointsPerMesh;
   numDecisionParams        = copy.numDecisionParams;
   numStateParams           = copy.numStateParams;
   numControlParams         = copy.numControlParams;
   numIntegralParams        = copy.numIntegralParams;
   numStaticParams          = copy.numStaticParams;
   integralStartIdx         = copy.integralStartIdx;
   integralStopIdx          = copy.integralStopIdx;
   timeStartIdx             = copy.timeStartIdx;
   timeStopIdx              = copy.timeStopIdx;
   staticStartIdx           = copy.staticStartIdx;
   staticStopIdx            = copy.staticStopIdx;
   
   Integer sz = (Integer) copy.decisionVector.GetSize();
   decisionVector.SetSize(sz);
   for (Integer ii = 0; ii < sz; ii++)
      decisionVector[ii] = copy.decisionVector[ii];

   return *this;
}


//------------------------------------------------------------------------------
// destructor
//------------------------------------------------------------------------------
DecisionVector::~DecisionVector()
{
   // nothing to do here
}

//------------------------------------------------------------------------------
//  void Initialize(Integer nStateVars,
//                  Integer nControlVars,
//                  Integer nIntegralParams,
//                  Integer nStaticParams,
//                  Integer nStateMeshPoints,
//                  Integer nControlMeshPoints,
//                  Integer nStateStagePoints,
//                  Integer nControlStagePoints)
//------------------------------------------------------------------------------
/**
 * This method initializes the decision vector
 *
 * @param <nStateVars>          number of state variables
 * @param <nControlVars>        number of control variables
 * @param <nIntegralParams>     number of integral parameters
 * @param <nStaticParams>       number of static parameters
 * @param <nStateMeshPoints>    number of state mesh points
 * @param <nControlMeshPoints>  number of control mesh points
 * @param <nStateStagePoints>   number of state stage points
 * @param <nControlStagePoints> number of control stage points
 *
 *
 */
//------------------------------------------------------------------------------
void DecisionVector::Initialize(Integer nStateVars,
                                Integer nControlVars,
                                Integer nIntegralParams,
                                Integer nStaticParams,
                                Integer nStateMeshPoints,
                                Integer nControlMeshPoints,
                                Integer nStateStagePoints,
                                Integer nControlStagePoints)
{
   #ifdef DEBUG_INIT
      MessageInterface::ShowMessage("In Init ...\n");
   #endif
   // Initialize the decision vector
   
   //  Check that inputs are valid
   if (nStateVars <= 0)
   {
      std::string errmsg = "For DecisionVector, numStateVars ";
      errmsg += "must be greater than 0.\n";
      throw LowThrustException(errmsg);
   }
   if (nStateMeshPoints <= 0)
   {
      std::string errmsg = "For DecisionVector, numStateMeshPoints ";
      errmsg += "must be greater than 0.\n";
      throw LowThrustException(errmsg);
   }
   if (nControlMeshPoints <= 0)
   {
      std::string errmsg = "For DecisionVector, numControlMeshPoints ";
      errmsg += "must be greater than 0.\n";
      throw LowThrustException(errmsg);
   }
   
   // Set optimal control problem sizes
   numStateVars          = nStateVars;
   numControlVars        = nControlVars;
   numIntegralParams     = nIntegralParams;
   numStaticParams       = nStaticParams;
   
   // Set discretization properites
   numStateMeshPoints    = nStateMeshPoints;
   numControlMeshPoints  = nControlMeshPoints;
   numStateStagePoints   = nStateStagePoints;
   numControlStagePoints = nControlStagePoints;
   
   SetChunkIndeces();
   #ifdef DEBUG_INIT
      MessageInterface::ShowMessage(
             "In Init ... setting size of decision vector to %d\n",
             numDecisionParams);
   #endif
   decisionVector.SetSize(numDecisionParams);
   decisionVector.MakeZeroVector();
}

//------------------------------------------------------------------------------
//  bool SetDecisionVector(const Rvector &decVector)
//------------------------------------------------------------------------------
/**
 * This method sets the decision vector to the input value
 *
 * @param <decVector>          decision vector
 *
 * @return true if successful; false, otherwise
 *
 */
//------------------------------------------------------------------------------
bool DecisionVector::SetDecisionVector(const Rvector &decVector)
{
   //  Set the entire decision vector
   
   // Check that decision vector is the required length
   Integer sz = decisionVector.GetSize();
   if (sz != numDecisionParams)
   {
      std::string errmsg = "For DecisionVector::SetDecisionVector, ";
      errmsg += "state vector is not valid length\n";
      throw LowThrustException(errmsg);
   }
   
   // Set the decision vector
   for (Integer ii = 0; ii < sz; ii++)
      decisionVector[ii]    = decVector[ii];

   return true;
}

//------------------------------------------------------------------------------
//  Rvector GetDecisionVector()
//------------------------------------------------------------------------------
/**
 * This method returns the decision vector value
 *
 * @return the current decision vector
 *
 */
//------------------------------------------------------------------------------
Rvector DecisionVector::GetDecisionVector()
{
   #ifdef DEBUG_GET
      MessageInterface::ShowMessage("GetDecisionVector:\n");
      MessageInterface::ShowMessage("   size of decision vector = %d\n",
                                    decisionVector.GetSize());
   #endif
   return decisionVector;
}


//------------------------------------------------------------------------------
//  const Rvector* GetDecisionVectorPointer()
//------------------------------------------------------------------------------
/**
 * This method returns a pointer to the decision vector
 *
 * @return a pointer to the decision vector
 *
 */
//------------------------------------------------------------------------------
const Rvector* DecisionVector::GetDecisionVectorPointer()
{
   return &decisionVector;
}

//------------------------------------------------------------------------------
//  bool SetTimeVector(const Rvector &tVector)
//------------------------------------------------------------------------------
/**
 * This method sets the time vector
 *
 * @param <tVector> the input time vector
 *
 * @return true if successful; false, otherwise
 *
 */
//------------------------------------------------------------------------------
bool DecisionVector::SetTimeVector(const Rvector &tVector)
{
   #ifdef DEBUG_SET
      MessageInterface::ShowMessage("SetTimeVector:\n");
      MessageInterface::ShowMessage("   timeStartIdx = %d\n", timeStartIdx);
      MessageInterface::ShowMessage("   timeStopIdx  = %d\n", timeStopIdx);
   #endif
   // Given a time vector, insert it in the decision vector
   
   if (tVector.GetSize() != 2)
   {
      std::string errmsg = "For DecisionVector::SetTimeVector, ";
      errmsg += "time vector is not valid dimension\n";
      throw LowThrustException(errmsg);
   }

   decisionVector[timeStartIdx] = tVector[0];
   decisionVector[timeStopIdx]  = tVector[1];
   
   return true;
}


//------------------------------------------------------------------------------
//  Rvector GetTimeVector()
//------------------------------------------------------------------------------
/**
 * This method returns the time vector
 *
 * @return the time vector
 *
 */
//------------------------------------------------------------------------------
Rvector DecisionVector::GetTimeVector()
{
   Rvector timeVec(2,decisionVector[timeStartIdx], decisionVector[timeStopIdx]);
   
   return timeVec;
}

//------------------------------------------------------------------------------
//  bool SetStaticVector(const Rvector &sVector)
//------------------------------------------------------------------------------
/**
 * This method sets the static vector
 *
 * @param <sVector> the static vector
 *
 * @return true if successful; false, otherwise
 *
 */
//------------------------------------------------------------------------------
bool DecisionVector::SetStaticVector(const Rvector &sVector)
{
   // Given a static vector, insert it in the decision vector
   
   Integer sz = sVector.GetSize(); // check against start-stop too?
   if (sz != numStaticParams)
   {
      std::string errmsg = "For DecisionVector::SetStaticVector, ";
      errmsg += "static vector is not valid dimension\n";
      throw LowThrustException(errmsg);
   }
   for (Integer ii = 0; ii < sz; ii++)
      decisionVector[staticStartIdx +ii] = sVector[ii];
   return true;
}

//------------------------------------------------------------------------------
//  Rvector  GetStaticVector()
//------------------------------------------------------------------------------
/**
 * This method returns the static vector
 *
 * @return the static vector
 *
 */
//------------------------------------------------------------------------------
Rvector DecisionVector::GetStaticVector()
{
   Rvector sVector(numStaticParams);
   for (Integer ii = 0; ii < numStaticParams; ii++)
      sVector[ii] = decisionVector[staticStartIdx + ii];
   
   return sVector;
}

//------------------------------------------------------------------------------
//  bool SetIntegralVector(const Rvector &iVector)
//------------------------------------------------------------------------------
/**
 * This method sets the integral vector
 *
 * @param <iVector> the integral vector
 *
 * @return true if successful; false, otherwise
 *
 */
//------------------------------------------------------------------------------
bool DecisionVector::SetIntegralVector(const Rvector &iVector)
{
   // Given an integral vector, insert it in the decision vector
   
   Integer sz = iVector.GetSize();
   if (sz != numIntegralParams)
   {
      std::string errmsg = "For DecisionVector::SetIntegralVector, ";
      errmsg += "integral vector is not valid dimension\n";
      throw LowThrustException(errmsg);
   }
   for (Integer ii = 0; ii < sz; ii++)
      decisionVector[integralStartIdx +ii] = iVector[ii];
   return true;
}


//------------------------------------------------------------------------------
//  Rvector GetIntegralVector()
//------------------------------------------------------------------------------
/**
 * This method returns the integral vector
 *
 * @return the integral vector
 *
 */
//------------------------------------------------------------------------------
Rvector DecisionVector::GetIntegralVector()
{
   Rvector iVector(numIntegralParams);
   for (Integer ii = 0; ii < numIntegralParams; ii++)
      iVector[ii] = decisionVector[integralStartIdx + ii];
   
   return iVector;
}

//------------------------------------------------------------------------------
//  Rvector GetFirstStateVector()
//------------------------------------------------------------------------------
/**
 * This method returns the first state vector
 *
 * @return the first state vector
 *
 */
//------------------------------------------------------------------------------
Rvector DecisionVector::GetFirstStateVector()
{
   return GetStateAtMeshPoint(0);
}

//------------------------------------------------------------------------------
//  Rvector GetLastStateVector()
//------------------------------------------------------------------------------
/**
 * This method returns the last state vector
 *
 * @return the last state vector
 *
 */
//------------------------------------------------------------------------------
Rvector DecisionVector::GetLastStateVector()
{
   return GetStateAtMeshPoint(numStateMeshPoints-1);
}

//------------------------------------------------------------------------------
//  Integer GetNumDecisionParams()
//------------------------------------------------------------------------------
/**
 * This method returns the number of decision parameters
 *
 * @return the number of decision parameters
 *
 */
//------------------------------------------------------------------------------
Integer DecisionVector::GetNumDecisionParams()
{
   return numDecisionParams;
}

//------------------------------------------------------------------------------
//  Integer GetNumStatePoints()
//------------------------------------------------------------------------------
/**
 * This method returns the number of state points
 *
 * @return the number of state points
 *
 */
//------------------------------------------------------------------------------
Integer DecisionVector::GetNumStatePoints()
{
   return numStatePoints;
}

//------------------------------------------------------------------------------
//  Integer GetNumControlPoints()
//------------------------------------------------------------------------------
/**
 * This method returns the number of control points
 *
 * @return the number of control points
 *
 */
//------------------------------------------------------------------------------
Integer DecisionVector::GetNumControlPoints()
{
   return numControlPoints;
}

//------------------------------------------------------------------------------
//  Integer GetNumStateVars()
//------------------------------------------------------------------------------
/**
 * This method returns the number of state variables
 *
 * @return the number of state variables
 *
 */
//------------------------------------------------------------------------------
Integer DecisionVector::GetNumStateVars()
{
   return numStateVars;
}

//------------------------------------------------------------------------------
//  Integer GetNumControlVars()
//------------------------------------------------------------------------------
/**
 * This method returns the number of control variables
 *
 * @return the number of control variables
 *
 */
//------------------------------------------------------------------------------
Integer DecisionVector::GetNumControlVars()
{
   return numControlVars;
}

//------------------------------------------------------------------------------
//  Real GetFirstTime()
//------------------------------------------------------------------------------
/**
 * This method returns the first time
 *
 * @return the first time
 *
 */
//------------------------------------------------------------------------------
Real DecisionVector::GetFirstTime()
{
   return decisionVector[timeStartIdx];
}

//------------------------------------------------------------------------------
//  Real GetLastTime()
//------------------------------------------------------------------------------
/**
 * This method returns the last time
 *
 * @return the last time
 *
 */
//------------------------------------------------------------------------------
Real DecisionVector::GetLastTime()
{
   return decisionVector[timeStopIdx];
}


//------------------------------------------------------------------------------
//  IntegerArray GetTimeIdxs()
//------------------------------------------------------------------------------
/**
 * This method returns the array of time indexes
 *
 * @return array of time indexes
 *
 */
//------------------------------------------------------------------------------
IntegerArray DecisionVector::GetTimeIdxs()
{
   IntegerArray idxs;
   for (Integer ii = timeStartIdx; ii <= timeStopIdx; ii++)
      idxs.push_back(ii);
   return idxs;
}



//------------------------------------------------------------------------------
//  Rvector GetStateAtMeshPoint(Integer meshIdx, Integer stageIdx = 0)
//------------------------------------------------------------------------------
/**
 * This method returns the state at the given mesh point and stage point
 *
 * @param <meshIdx>   input mesh point
 * @param <stageIdx>  input stage point
 *
 * @return state at the given mesh and stage point
 *
 */
//------------------------------------------------------------------------------
Rvector DecisionVector::GetStateAtMeshPoint(Integer meshIdx, Integer stageIdx)
{
   // Check that state vector contains the mesh point
   if (meshIdx < 0 || meshIdx > numStatePoints)
   {
      std::string errmsg = "For DecisionVector::GetStateAtMeshPoint, ";
      errmsg += "meshIdx must be >= 0 and <= numStateMeshPoints\n";
      throw LowThrustException(errmsg);
   }
   
   // Compute indeces of state vector components then extract the
   // state vector
   IntegerArray idxs = GetStateIdxsAtMeshPoint(meshIdx, stageIdx);
   #ifdef DEBUG_DEC_VEC_STATE
      MessageInterface::ShowMessage("In GetStateAtMeshPoint, idxs size = %d\n",
                                    idxs.size());
      MessageInterface::ShowMessage(
                        "In GetStateAtMeshPoint, decisionVector size = %d\n",
                        decisionVector.GetSize());
   #endif
   Integer num = idxs.size();
   Rvector theStateVector(num);
   for (Integer ii = 0; ii < num; ii++)
   {
      #ifdef DEBUG_DEC_VEC_STATE
         MessageInterface::ShowMessage(
                        "                  decisionVector(%d) = %12.10f\n",
                        idxs[ii], decisionVector[idxs[ii]]);
      #endif
      theStateVector[ii] = decisionVector[idxs[ii]];
   }
   
   return theStateVector;
}

//------------------------------------------------------------------------------
//  Rvector GetControlAtMeshPoint(Integer meshIdx, Integer stageIdx = 0)
//------------------------------------------------------------------------------
/**
 * This method returns the control at the given mesh point and stage point
 *
 * @param <meshIdx>   input mesh point
 * @param <stageIdx>  input stage point
 *
 * @return control at the given mesh and stage point
 *
 */
//------------------------------------------------------------------------------
Rvector DecisionVector::GetControlAtMeshPoint(Integer meshIdx, Integer stageIdx)
{
   // Check that state vector contains the mesh point
   if (meshIdx < 0 || meshIdx > numControlPoints)
   {
      std::string errmsg = "For DecisionVector::GetControlAtMeshPoint, ";
      errmsg += "meshIdx must be >= 0 and <= numControlMeshPoints\n";
      throw LowThrustException(errmsg);
   }
   
   // Compute indeces of control vector components the extract the
   // control vector
   IntegerArray idxs = GetControlIdxsAtMeshPoint(meshIdx, stageIdx);
   Integer num = idxs.size();
   Rvector theControlVector(num);
   for (Integer ii = 0; ii < num; ii++)
      theControlVector[ii] = decisionVector[idxs[ii]];
   
   return theControlVector;
}

//------------------------------------------------------------------------------
//  Rvector  GetInterpolatedStateVector(Real atTime)
//------------------------------------------------------------------------------
/**
 * This method returns the interpolated state vector at the given time
 *
 * @param <atTime>   input time
 *
 * @return interpolated state vector at the given time
 *
 */
//------------------------------------------------------------------------------
Rvector DecisionVector::GetInterpolatedStateVector(Real atTime)
{
   // NOTE THIS IS A DUMMY IMPLEMENTATION THAT CURRENTLY ONLY TESTS
   // THAT THE INTERFACES ARE WORKING
   //   warning('DecisionVector::GetInterpolatedStateVector is not implemented')
   Rvector stateVec(numStateVars); // default is all zeroes
   return stateVec;
}


//------------------------------------------------------------------------------
//  Rvector  GetInterpolatedControlVector(Real atTime)
//------------------------------------------------------------------------------
/**
 * This method returns the interpolated control vector at the given time
 *
 * @param <atTime>   input time
 *
 * @return interpolated control vector at the given time
 *
 */
//------------------------------------------------------------------------------
Rvector DecisionVector::GetInterpolatedControlVector(Real atTime)
{
   // NOTE THIS IS A DUMMY IMPLEMENTATION THAT CURRENTLY ONLY TESTS
   // THAT THE INTERFACES ARE WORKING
   //   warning('DecisionVector::GetInterpolatedStateVector is not implemented')
   Rvector controlVec(numControlVars); // default is all zeroes
   return controlVec;
}

